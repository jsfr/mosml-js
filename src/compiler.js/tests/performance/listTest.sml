prim_val ctimebegin_ : string -> unit = 1 "console.time";
prim_val ctimeend_ : string -> unit = 1 "console.timeEnd";
fun ctimeBegin n = ctimebegin_ n;
fun ctimeEnd n = ctimeend_ n;

prim_val clog_ : 'a -> unit = 1 "console.log";
fun clog x = clog_ x;
fun length xs =
    let fun acc []      k = k
          | acc (x::xr) k = acc xr (k+1)
    in acc xs 0 end;


local
  fun append [] ys = ys
    | append (x::xs) ys = x :: append xs ys
in
  fun xs @ [] = xs
    | xs @ ys = append xs ys
end


fun repeat n f x =
  let 
    fun loop 0 = f x
      | loop n = (f x; loop(n-1))
  in 
    loop(n-1) 
  end 
fun timerep n f s = repeat n (fn x => (ctimeBegin(s);f x; ctimeEnd(s)))


fun listPrepend x =
  let
    val addRef = ref 0.0;
    val listRef = ref []
  in
    while !addRef < x do 
      (addRef := !addRef+1.0;
       listRef := 1::(!listRef)
      )
  end

fun listAppend x =
  let
    val addRef = ref 0.0;
    val listRef = ref []
  in
    while !addRef < x do 
      (addRef := !addRef+1.0;
       listRef := (!listRef)@[1]
      )
  end


val it = timerep 10 listPrepend "listPrepend" 100000.0
val it = timerep 10 listAppend "listPrepend" 1000.0


