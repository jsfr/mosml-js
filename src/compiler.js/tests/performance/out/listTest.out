beautified listTest.js
==============================JS BEGIN==============================
listPrepend: 72ms
listPrepend: 63ms
listPrepend: 74ms
listPrepend: 73ms
listPrepend: 76ms
listPrepend: 77ms
listPrepend: 76ms
listPrepend: 78ms
listPrepend: 74ms
listPrepend: 90ms
listPrepend: 74ms
listPrepend: 78ms
listPrepend: 81ms
listPrepend: 74ms
listPrepend: 78ms
listPrepend: 67ms
listPrepend: 68ms
listPrepend: 68ms
listPrepend: 69ms
listPrepend: 81ms
listPrepend: 79ms
listPrepend: 82ms
listPrepend: 76ms
listPrepend: 74ms
listPrepend: 78ms
listPrepend: 66ms
listPrepend: 62ms
listPrepend: 62ms
listPrepend: 71ms
listPrepend: 74ms
listPrepend: 74ms
listPrepend: 74ms
listPrepend: 74ms
listPrepend: 77ms
listPrepend: 72ms
listPrepend: 79ms
listPrepend: 99ms
listPrepend: 81ms
listPrepend: 84ms
listPrepend: 76ms
listPrepend: 77ms
listPrepend: 71ms
listPrepend: 81ms
listPrepend: 69ms
listPrepend: 72ms
listPrepend: 77ms
listPrepend: 77ms
listPrepend: 79ms
listPrepend: 71ms
listPrepend: 64ms
listPrepend: 72ms
listPrepend: 74ms
listPrepend: 73ms
listPrepend: 75ms
listPrepend: 73ms
listPrepend: 78ms
listPrepend: 66ms
listPrepend: 62ms
listPrepend: 62ms
listPrepend: 63ms
listPrepend: 72ms
listPrepend: 75ms
listPrepend: 74ms
listPrepend: 79ms
listPrepend: 75ms
listPrepend: 85ms
listPrepend: 71ms
listPrepend: 72ms
listPrepend: 75ms
listPrepend: 72ms
listPrepend: 77ms
listPrepend: 72ms
listPrepend: 79ms
listPrepend: 66ms
listPrepend: 73ms
listPrepend: 74ms
listPrepend: 73ms
listPrepend: 76ms
listPrepend: 74ms
listPrepend: 78ms
listPrepend: 72ms
listPrepend: 81ms
listPrepend: 68ms
listPrepend: 71ms
listPrepend: 78ms
listPrepend: 73ms
listPrepend: 81ms
listPrepend: 67ms
listPrepend: 72ms
listPrepend: 74ms
listPrepend: 77ms
listPrepend: 78ms
listPrepend: 78ms
listPrepend: 81ms
listPrepend: 72ms
listPrepend: 87ms
listPrepend: 67ms
listPrepend: 71ms
listPrepend: 79ms
listPrepend: 74ms
listAppend: 170ms
listAppend: 152ms
listAppend: 153ms
listAppend: 151ms
listAppend: 152ms
listAppend: 152ms
listAppend: 152ms
listAppend: 152ms
listAppend: 180ms
listAppend: 156ms
listAppend: 151ms
listAppend: 151ms
listAppend: 153ms
listAppend: 151ms
listAppend: 152ms
listAppend: 151ms
listAppend: 151ms
listAppend: 152ms
listAppend: 152ms
listAppend: 152ms
listAppend: 152ms
listAppend: 159ms
listAppend: 159ms
listAppend: 151ms
listAppend: 151ms
listAppend: 151ms
listAppend: 151ms
listAppend: 153ms
listAppend: 152ms
listAppend: 152ms
listAppend: 152ms
listAppend: 152ms
listAppend: 152ms
listAppend: 151ms
listAppend: 157ms
listAppend: 165ms
listAppend: 156ms
listAppend: 152ms
listAppend: 152ms
listAppend: 151ms
listAppend: 151ms
listAppend: 152ms
listAppend: 151ms
listAppend: 152ms
listAppend: 152ms
listAppend: 154ms
listAppend: 152ms
listAppend: 156ms
listAppend: 159ms
listAppend: 151ms
listAppend: 151ms
listAppend: 152ms
listAppend: 150ms
listAppend: 152ms
listAppend: 152ms
listAppend: 151ms
listAppend: 151ms
listAppend: 150ms
listAppend: 152ms
listAppend: 152ms
listAppend: 156ms
listAppend: 160ms
listAppend: 151ms
listAppend: 151ms
listAppend: 152ms
listAppend: 151ms
listAppend: 151ms
listAppend: 152ms
listAppend: 151ms
listAppend: 151ms
listAppend: 152ms
listAppend: 152ms
listAppend: 151ms
listAppend: 156ms
listAppend: 160ms
listAppend: 151ms
listAppend: 151ms
listAppend: 151ms
listAppend: 152ms
listAppend: 150ms
listAppend: 152ms
listAppend: 152ms
listAppend: 154ms
listAppend: 151ms
listAppend: 152ms
listAppend: 151ms
listAppend: 156ms
listAppend: 161ms
listAppend: 151ms
listAppend: 152ms
listAppend: 151ms
listAppend: 152ms
listAppend: 152ms
listAppend: 153ms
listAppend: 152ms
listAppend: 151ms
listAppend: 152ms
listAppend: 152ms
listAppend: 153ms
listAppend: 157ms
==============================JS END================================
